kernel/reduction.ml

```diff
@@ -353,9 +353,89 @@ let irr_flex env = function
   | VarKey x -> Context.Named.Declaration.get_relevance (Environ.lookup_named x env) == Sorts.Irrelevant
   | RelKey x -> Context.Rel.Declaration.get_relevance (Environ.lookup_rel x env) == Sorts.Irrelevant
 
+let leq_universes univs (_,e1) (_,e2) u1 u2 =
+  let subst e u = if Univ.Instance.is_empty e then u else Univ.subst_instance_instance e u in
+  UGraph.check_eq_instances univs (subst e1 u1) (subst e2 u2)
+
+let rec compare_under univs pb e1 c1 e2 c2 =
+  match Constr.kind c1, Constr.kind c2 with
+  | Cast (c1, _, _), _ -> compare_under univs pb e1 c1 e2 c2
+  | _, Cast (c2, _, _) -> compare_under univs pb e1 c1 e2 c2
+  | Rel i, Rel j -> begin match Esubst.expand_rel i (fst e1) with
+      | Inl _ -> false
+      | Inr (k, _) -> begin match Esubst.expand_rel j (fst e2) with
+          | Inl _ -> false
+          | Inr (k', _) -> Int.equal k k'
+        end
+    end
+  | Meta m1, Meta m2 -> Int.equal m1 m2
+  | Var id1, Var id2 -> Id.equal id1 id2
+  | Int i1, Int i2 -> Uint63.equal i1 i2
+  | Float f1, Float f2 -> Float64.equal f1 f2
+  | Sort s1, Sort s2 ->
+    let subst_instance_sort u s =
+      if Univ.Instance.is_empty u then s else Sorts.subst_instance_sort u s
+    in
+    begin match pb with
+    | CONV ->
+      UGraph.check_eq_sort univs
+        (subst_instance_sort (snd e1) s1)
+        (subst_instance_sort (snd e2) s2)
+    | CUMUL ->
+      UGraph.check_leq_sort univs
+        (subst_instance_sort (snd e1) s1)
+        (subst_instance_sort (snd e2) s2)
+    end
+  | Prod (_,t1,c1), Prod (_,t2,c2) ->
+    compare_under univs CONV e1 t1 e2 t2
+    && compare_under univs CUMUL (usubs_lift e1) c1 (usubs_lift e2) c2
```

Comment:
This `CUMUL` should read `pb` here.

Reply:
OK, so please remove it so that we only care about trivial equality comparisons.

Reply:
changed